rpc/rpc-lib/src/rpc-clnt.c

```diff
@@ -1696,41 +1699,50 @@ rpc_clnt_submit(struct rpc_clnt *rpc, rpc_clnt_prog_t *prog, int procnum,
                        conn->config.remote_host, conn->config.remote_port);
                 goto unlock;
             }
+            lock_flag = _gf_false;
         }
-
-        ret = rpc_transport_submit_request(conn->trans, &req);
-        if (ret == -1) {
-            gf_log(conn->name, GF_LOG_WARNING,
-                   "failed to submit rpc-request "
-                   "(unique: %" PRIu64
-                   ", XID: 0x%x Program: %s, "
-                   "ProgVers: %d, Proc: %d) to rpc-transport (%s)",
-                   cframe->root->unique, rpcreq->xid, rpcreq->prog->progname,
-                   rpcreq->prog->progver, rpcreq->procnum, conn->name);
-        } else if ((ret >= 0) && frame) {
-            /* Save the frame in queue */
-            __save_frame(rpc, frame, rpcreq);
-
-            /* A ref on rpc-clnt object is taken while registering
-             * call_bail to timer in __save_frame. If it fails to
-             * register, it needs an unref and should happen outside
-             * conn->lock which otherwise leads to deadlocks */
+        if (frame)
+            save_frame = __save_frame(rpc, frame, rpcreq);
+    }
+    pthread_mutex_unlock(&conn->lock);
+    ret = rpc_transport_submit_request(conn->trans, &req);
+    if (ret == -1) {
+        pthread_mutex_lock(&conn->lock);
+        {
+            list_del_init(&save_frame->list);
+        }
+        pthread_mutex_unlock(&conn->lock);
+        gf_log(conn->name, GF_LOG_WARNING,
+               "failed to submit rpc-request "
+               "(unique: %" PRIu64
+               ", XID: 0x%x Program: %s, "
+               "ProgVers: %d, Proc: %d) to rpc-transport (%s)",
+               cframe->root->unique, rpcreq->xid, rpcreq->prog->progname,
```

Comment:
question: without locking, how do you know the values you fetch here from rpcreq are valid?

Reply:
The object rpcreq is prepared locally and we are passing the information that is coming from parent function so i don't believe we need to take any lock to print the rpcreq details.

Reply:
conn->name is not local (and btw is used twice - first as the domain and then in the message)

Reply:
Yes, i will change the same in next patch.